fe/fe-core/src/main/java/org/apache/doris/nereids/memo/Memo.java

```diff
@@ -706,4 +710,172 @@ public String toString() {
+        int prefix = res.size();
+        for (GroupExpression groupExpression : extractGroupExpressionContainsProp(group, prop)) {
+            for (Pair<Long, Double> idCostPair : rankGroupExpression(groupExpression, prop)) {
+                res.add(Pair.of(idCostPair.first + prefix, idCostPair.second));
+            }
+            prefix = res.size();
+        }
+        return res;
+    }
+
+    private List<Pair<Long, Double>> rankGroupExpression(GroupExpression groupExpression,
+            PhysicalProperties prop) {
+        if (!groupExpression.getLowestCostTable().containsKey(prop)) {
+            return new ArrayList<>();
+        }
+        List<Pair<Long, Double>> res = new ArrayList<>();
+
+        List<PhysicalProperties> inputProperties = groupExpression.getInputPropertiesList(prop);
+        if (groupExpression.getPlan() instanceof LeafPlan) {
+            res.add(Pair.of(0L, groupExpression.getCostByProperties(prop)));
+            return res;
+        }
+
+        double bestChildrenCost = 0;
+        List<List<Pair<Long, Double>>> children = new ArrayList<>();
+        for (int i = 0; i < inputProperties.size(); i++) {
+            Preconditions.checkArgument(!groupExpression.child(i).equals(groupExpression.getOwnerGroup())
+                    || !prop.equals(inputProperties.get(i)));
+            bestChildrenCost += groupExpression.children().get(i).getLowestCostPlan(inputProperties.get(i)).get().first;
+            List<Pair<Long, Double>> idCostPair
+                    = rankGroup(groupExpression.child(i), inputProperties.get(i));
+            children.add(idCostPair);
+        }
+        List<Pair<Long, List<Integer>>> childrenId = new ArrayList<>();
+        permute(children, 0, childrenId, new ArrayList<>());
+        for (Pair<Long, List<Integer>> c : childrenId) {
+            double childCost = 0;
+            for (int i = 0; i < children.size(); i++) {
+                childCost += children.get(i).get(c.second.get(i)).second;
+            }
+            res.add(Pair.of(c.first,
+                    childCost + groupExpression.getCostByProperties(prop) - bestChildrenCost));
+        }
+        return res;
+    }
+
+    // we permute all children, e.g.,
+    //      for children [1, 2] [1, 2, 3]
+    //      we can get: 0: [1,1] 1:[1, 2] 2:[1, 3] 3:[2, 1] 4:[2, 2] 5:[2, 3]
+    private void permute(List<List<Pair<Long, Double>>> children, int index,
+            List<Pair<Long, List<Integer>>> result, List<Integer> current) {
+        if (index == children.size()) {
+            result.add(Pair.of(getUniqueId(children, current), current));
+            return;
+        }
+        for (int i = 0; i < children.get(index).size(); i++) {
+            List<Integer> next = new ArrayList<>(current);
+            next.add(i);
+            permute(children, index + 1, result, next);
+        }
+    }
+
+    private static long getUniqueId(List<List<Pair<Long, Double>>> lists, List<Integer> current) {
```

Comment:
add some comment to explain the logic of id generator

Reply:
done